### PR TITLE
[indexer] handle token create, mint, and direct transfers

### PR DESCRIPTION
**Pull request by bowenyang007**
#### TL;DR
The transaction looks like this for direct transfers, create and mint tokens: 
```
write set changes 
   write_resource
        address -> TokenStore<handle>
   write_table_item
        handle
              TokenId -> Token
```
We can easily parse out the token and then find the owner via write_resources. This however doesn't apply to claim/offer and swap token for coin contracts. 

#### Details

We're using write-set to get the final state in each transaction. Table items will have the token metadata and the resources will have the owner of the tables. We then connect token metadata to the owner via table handle. 

In the case of offer / claim, we won't be able to find the owner via this methodology so the plan right now is to just leave owner address as null. We'll still index the token so that we know that something happened, but we won't know who owns it until we handle those special cases separately. 

### Test Plan
Performed end to end testing with indexer connected to staging firehose. 

Running block_to_block_output
```
> DATABASE_URL=postgres://postgres@localhost:5432/aptos_indexer  cargo run -- --endpoint-url http://localhost:9000 --package-file ../aptos-substreams/aptos-substreams-v0.0.3.spkg --module-name block_to_block_output --emit-every 50
```

Running block_output_to_token
```
> INSPECTION_PORT=9106 DATABASE_URL=postgres://postgres@localhost:5432/aptos_indexer  cargo run -- --endpoint-url http://localhost:9000 --package-file ../aptos-substreams/aptos-substreams-v0.0.3.spkg --module-name block_output_to_token --emit-every 50
```

Then verified that db works. 

- - -
This change is [Reviewable](https://reviewable.io/reviews/aptos-labs/aptos-core/3873)

### CODE REVIEW

File: crates/aptos-protos/proto/aptos/tokens/v1/tokens.proto [hunk 1]
```diff
@@ -0,0 +1,65 @@
+// Copyright (c) Aptos
+// SPDX-License-Identifier: Apache-2.0
+
+syntax = "proto3";
+
+package aptos.tokens.v1;
+
+
+message Tokens {
+  uint64 block_height = 1;
```
Comment (areshand) [inline crates/aptos-protos/proto/aptos/tokens/v1/tokens.proto:10]: what is block_height?

File: crates/aptos-protos/proto/aptos/tokens/v1/tokens.proto [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+  uint32 chain_id = 2;
+  repeated Token tokens = 3;
+  repeated TokenData token_datas = 4;
+  repeated CollectionData collection_datas = 5;
+}
+
+message Token {
+  TokenId token_id = 1;
+  uint64 transaction_version = 2;
```
Comment (areshand) [inline crates/aptos-protos/proto/aptos/tokens/v1/tokens.proto:19]: just curious if we modify the token's amount what would be the transaction_version?

File: ecosystem/sf-indexer/aptos-sf-indexer/src/models/tokens.rs [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+            .token_data_id
+            .as_ref()
+            .unwrap()
+            .clone();
+        Self {
+            creator_address: token_data_id.creator_address,
+            collection_name: token_data_id.collection_name,
+            name: token_data_id.name,
+            property_version: u64_to_bigdecimal(token.token_id.as_ref().unwrap().property_version),
+            transaction_version: token.transaction_version as i64,
```
Comment (areshand) [inline ecosystem/sf-indexer/aptos-sf-indexer/src/models/tokens.rs:171]: with tnx_version, to get the owned amount seems requires one query to get the largest version and another query to get corresponding amount now. Is this intented?

File: crates/aptos-protos/proto/aptos/tokens/v1/tokens.proto (hunk 1 above)
Comment (bowenyang007) [reply crates/aptos-protos/proto/aptos/tokens/v1/tokens.proto:10]: It's the block number.

File: ecosystem/sf-indexer/aptos-sf-indexer/src/models/tokens.rs (hunk 3 above)
Comment (bowenyang007) [reply ecosystem/sf-indexer/aptos-sf-indexer/src/models/tokens.rs:171]: ah yes. we'll have a current version of this table. This table is for transaction history.

File: crates/aptos-protos/proto/aptos/tokens/v1/tokens.proto (hunk 2 above)
Comment (bowenyang007) [reply crates/aptos-protos/proto/aptos/tokens/v1/tokens.proto:19]: Each transaction version will be recorded as a separate row. So we'd have 2 rows w/ different token amounts.

File: ecosystem/sf-indexer/aptos-substreams/src/lib.rs [hunk 4]
```diff
@@ -85,26 +88,71 @@ fn block_to_block_output(input_block: Block) -> Result<BlockOutput, Error> {
...
+#[substreams::handlers::map]
+fn block_output_to_token(block: BlockOutput) -> Result<Tokens, Error> {
+    let mut tokens = vec![];
+    // let mut token_ownerships = vec![];
+    let mut token_datas = vec![];
+    let mut collection_datas = vec![];
+    for txn in block.transactions {
+        let txn_version = txn.transaction_info_output.unwrap().version;
+        // Claims will be supported separately
+        let supported_functions = vec![
```
Comment (areshand) [inline ecosystem/sf-indexer/aptos-substreams/src/lib.rs:100]: would we be able to index data from other modules calling our public functions? for example, we have a public function here https://github.com/aptos-labs/aptos-core/blob/7131fad47d92b893dadaa2125e4cae73f82e680d/aptos-move/framework/aptos-token/sources/token.move#L483. Other contracts can call to transfer token. In this case, the function is not the entry function.

Comment (bowenyang007) [reply ecosystem/sf-indexer/aptos-substreams/src/lib.rs]: Yup, I changed the logic. Basically if we somehow can't find the owner address we'll just index the rest of the token and leave owner address as missing. Rn it'd be missing for claim / offer for example. This way we can be a lot more generic.

File: ecosystem/sf-indexer/aptos-substreams/src/lib.rs [hunk 5]
```diff
@@ -85,26 +88,61 @@ fn block_to_block_output(input_block: Block) -> Result<BlockOutput, Error> {
...
-fn generate_trx_type_key(trx_type: TransactionType) -> String {
-    match trx_type {
-        TransactionType::Genesis => "genesis",
-        TransactionType::BlockMetadata => "block_metadata",
-        TransactionType::StateCheckpoint => "state_checkpoint",
-        TransactionType::User => "user",
+        if let Some(TxnDataOutput::User(_)) = &txn.txn_data {
+        } else {
+            continue;
+        }
```
Comment (gregnazario) [inline ecosystem/sf-indexer/aptos-substreams/src/lib.rs:103]: This should likely be a `if !matches(&txn.txn_data, Some(TxnDataOutput::User(_)))` statement, and I'm amazed that the linter didn't pick this up

File: ecosystem/sf-indexer/aptos-substreams/src/lib.rs (hunk 4 above)
Comment (areshand) [reply ecosystem/sf-indexer/aptos-substreams/src/lib.rs]: all the data from public function being called by other modules are filtered out right? there are other public function that could be called ex: this is to create more tokens https://github.com/aptos-labs/aptos-core/blob/8f8de4adcdaa0edbac2cb3c809c278d97331bdba/aptos-move/framework/aptos-token/sources/token.move#L751

File: ecosystem/sf-indexer/aptos-substreams/src/token_converter.rs [hunk 6]
```diff
@@ -0,0 +1,369 @@
...
+                    "version {} failed! mutability_config.uri missing {:?}",
+                    txn_version, value
+                ))?,
+            description_mutable: value["mutability_config"]["description"]
+                .as_bool()
+                .context(format!(
+                    "version {} failed! mutability_config.description missing {:?}",
+                    txn_version, value
+                ))?,
+        }));
```
Comment (gregnazario) [inline ecosystem/sf-indexer/aptos-substreams/src/token_converter.rs:366]: Btw we should probably use constants for things that are common here, this seems pretty fragile

File: ecosystem/sf-indexer/aptos-substreams/src/token_converter.rs [hunk 7]
```diff
@@ -0,0 +1,369 @@
...
+    }
+    Ok(None)
+}
+
+pub fn get_collection_data(
+    table_item: &TableItemOutput,
+    txn_version: u64,
+    table_handle_to_owner: &HashMap<String, (String, String)>,
```
Comment (gregnazario) [inline ecosystem/sf-indexer/aptos-substreams/src/token_converter.rs:291]: You should use a type to describe the three strings, for someone who looks at this, it's wondering what's the 2 strings at the end?  Especially since you throw away the second one a few lines below

File: ecosystem/sf-indexer/aptos-substreams/src/token_converter.rs [hunk 8]
```diff
@@ -0,0 +1,369 @@
...
+pub fn get_token(
+    table_item: &TableItemOutput,
+    txn_version: u64,
+    table_handle_to_owner: &HashMap<String, (String, String)>,
+) -> anyhow::Result<Option<Token>> {
+    if table_item.key_type != "0x3::token::TokenId" {
+        return Ok(None);
+    }
+    let table_handle = standardize_handle(&table_item.handle);
+    let (owner_address, table_type) = table_handle_to_owner
```
Comment (gregnazario) [inline ecosystem/sf-indexer/aptos-substreams/src/token_converter.rs:65]: Is this the table value type?

File: ecosystem/sf-indexer/aptos-substreams/src/token_converter.rs [hunk 9]
```diff
@@ -0,0 +1,369 @@
...
+                .map(|s| s.to_string())
+                .context(format!(
+                    "version {} failed! collection data handle must be present {:?}",
+                    txn_version, data
+                ))?;
+
+            Ok(Some(HashMap::from([(
+                standardize_handle(&collection_handle),
+                (address.clone(), resource.type_str.clone()),
+            )])))
```
Comment (gregnazario) [inline ecosystem/sf-indexer/aptos-substreams/src/token_converter.rs:31]: Can you make a struct for the 2nd part?

Also have a type alias for the first one?

**Review by gregnazario — COMMENTED** [review]
So my main feedback is that as an outside reader, I don't know what's going on.  Because I've worked with similar things, I know you're parsing the outputs to determine what the token is, but you should add some comments so if someone else has to look at it, they can fix it.

This needs more documentation

File: ecosystem/sf-indexer/aptos-sf-indexer/src/indexer/substream_processor.rs [hunk 10]
```diff
@@ -227,6 +227,19 @@ pub fn get_start_block(pool: &PgDbPool, substream_module_name: &String) -> Optio
                 substream_module = $1
                 AND success = TRUE
         ),
+        boundaries AS
+        (
+            SELECT
+                MAX(block_height) AS MAX_V,
+                MIN(block_height) AS MIN_V
```
[the remark this answers is not in the document]
Comment (bowenyang007) [reply ecosystem/sf-indexer/aptos-sf-indexer/src/indexer/substream_processor.rs:234]: oh it should be MAX_BLOCK. let me change that.

File: crates/aptos-protos/proto/aptos/block_output/v1/block_output.proto [hunk 11]
```diff
@@ -66,6 +66,7 @@ message UserTransactionOutput {
   aptos.util.timestamp.Timestamp timestamp = 8;
   repeated SignatureOutput signatures = 9;
   string payload = 10;
+  string entry_function_id_str = 11;
```
Comment (CapCap) [inline crates/aptos-protos/proto/aptos/block_output/v1/block_output.proto:69]: can this be a struct? is it a struct?

File: ecosystem/sf-indexer/aptos-substreams/src/token_converter.rs (hunk 9 above)
Comment (bowenyang007) [reply ecosystem/sf-indexer/aptos-substreams/src/token_converter.rs:31]: sure.

File: crates/aptos-protos/proto/aptos/block_output/v1/block_output.proto (hunk 11 above)
Comment (bowenyang007) [reply crates/aptos-protos/proto/aptos/block_output/v1/block_output.proto:69]: Actually I don't need this anymore let me remove it.

File: ecosystem/sf-indexer/aptos-substreams/src/token_converter.rs (hunk 8 above)
Comment (bowenyang007) [reply ecosystem/sf-indexer/aptos-substreams/src/token_converter.rs:65]: No it's the resource type and if the resource is a table i'm calling it table type.

File: ecosystem/sf-indexer/aptos-substreams/src/token_converter.rs (hunk 7 above)
Comment (bowenyang007) [reply ecosystem/sf-indexer/aptos-substreams/src/token_converter.rs:291]: Done.

File: ecosystem/sf-indexer/aptos-substreams/src/token_converter.rs (hunk 6 above)
Comment (bowenyang007) [reply ecosystem/sf-indexer/aptos-substreams/src/token_converter.rs:366]: Hmm yes but changing any of these would be a breaking change IMO.

File: ecosystem/sf-indexer/aptos-substreams/Makefile [hunk 12]
```diff
@@ -2,6 +2,10 @@
...
-stream:
-	substreams run -p -e localhost:18015 substreams.yaml block_to_block_output --start-block 0 --stop-block +100
\ No newline at end of file
+.PHONY: stream_block
+stream_block:
+	substreams run -p -e localhost:9000 substreams.yaml block_to_block_output --start-block 4466 --stop-block +0
+
+.PHONY: stream_token
+stream_token:
+	substreams run -p -e localhost:9000 substreams.yaml block_output_to_token --start-block 64205 --stop-block +0
```
[the remark this answers is not in the document]
Comment (bowenyang007) [reply ecosystem/sf-indexer/aptos-substreams/Makefile]: These scripts are just for testing.

File: ecosystem/sf-indexer/aptos-substreams/Cargo.lock [hunk 13]
```diff
@@ -22,6 +22,7 @@ dependencies = [
 name = "aptos-substreams"
 version = "0.0.1"
 dependencies = [
+ "anyhow",
```
[the remark this answers is not in the document]
Comment (bowenyang007) [reply ecosystem/sf-indexer/aptos-substreams/Cargo.lock:25]: B/c we have to exclude this from the main workspace to avoid building an overly large wasm which causes another set of problems.

File: ecosystem/sf-indexer/aptos-sf-indexer/src/indexer/substream_processor.rs [hunk 14]
```diff
@@ -269,6 +282,8 @@ pub fn get_start_block(pool: &PgDbPool, substream_module_name: &String) -> Optio
     }
     let mut res: Vec<Option<Gap>> = sql_query(sql)
         .bind::<Text, _>(substream_module_name)
+        // This is the number used to determine how far we look back for gaps. Increasing it may result in slower startup
+        .bind::<BigInt, _>(1500000)
         .get_results(&conn)
```
[the remark this answers is not in the document]
Comment (bowenyang007) [reply ecosystem/sf-indexer/aptos-sf-indexer/src/indexer/substream_processor.rs:287]: startup. I'll add comment.

File: ecosystem/sf-indexer/aptos-substreams/src/lib.rs (hunk 4 above)
Comment (bowenyang007) [reply ecosystem/sf-indexer/aptos-substreams/src/lib.rs]: hmm idk what your question is sorry. if the token is 0x3 i'm assuming that we should index it. is this not the right assumptino?

File: ecosystem/sf-indexer/aptos-sf-indexer/src/substream_processors/tokens_processor.rs [hunk 15]
```diff
@@ -0,0 +1,241 @@
...
+use aptos_protos::tokens::v1::Tokens;
+use async_trait::async_trait;
+use diesel::result::Error;
+use field_count::FieldCount;
+use prost::Message;
+use std::fmt::Debug;
+
+pub struct TokensSubstreamProcessor {
+    connection_pool: PgDbPool,
+    is_chain_id_verified: bool,
```
Comment (gregnazario) [inline ecosystem/sf-indexer/aptos-sf-indexer/src/substream_processors/tokens_processor.rs:25]: interesting, assuming this only really happens once

**Review by gregnazario — APPROVED** [review]

File: ecosystem/sf-indexer/aptos-substreams/src/lib.rs (hunk 4 above)
Comment (areshand) [reply ecosystem/sf-indexer/aptos-substreams/src/lib.rs]: I see the filter is deleted, which make senses.

**Review by areshand — APPROVED** [review]
